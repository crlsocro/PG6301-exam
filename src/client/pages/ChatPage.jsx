import * as React from "react";
import { useEffect, useState } from "react";
import * as ReactDOM from "react-dom";
import { ChatView } from "../views/ChatView";

export function ChatPage() {
    const [chatLog, setChatLog] = useState([]);
    const [ws, setWs] = useState();
    useEffect(() => {
        const ws = new WebSocket("ws://localhost:3000");
        ws.onopen = (event) => {
            console.log("opened", event);
        };
        ws.onmessage = (event) => {
            console.log("from server", event);
            setChatLog((chatLog) => [...chatLog, event.data]);
        };
        ws.onclose = (event) => {
            console.log("close", event);
        };
        setWs(ws);
    }, []);

    return (
        <ChatView chatLog={chatLog} onSendMessage={(message) => ws.send(message)} />
    );
}