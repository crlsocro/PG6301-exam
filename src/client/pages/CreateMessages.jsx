import React from "react";
import { Link } from "react-router-dom";
import {useLoading} from "../views/useLoading";
import {ErrorView} from "../views/ErrorView";
import {LoadingView} from "../views/LoadingView";


export function CreateMessages({userApi}) {
    async function submit(e) {
        e.preventDefault();
        await userApi.listUsers({firstName, lastName});
    }

    const { data: users, error, loading, reload } = useLoading(
        async () => await userApi.listUsers()
    );
    if (error) {
        return <ErrorView error={error} reload={reload} />;
    }

    if (loading || !users) {
        return <LoadingView />;
    }
    return (
        <>
            <h1>Send Message</h1>
            <form onSubmit={submit}>
            {users.map(({ id, firstName, lastName}) => (
                <select  key={id}>
                    <option value={`/users/${id}`}>{firstName} {lastName}</option>
                </select>
            ))}
                <button>Submit</button>
            </form>
        </>
    );
}