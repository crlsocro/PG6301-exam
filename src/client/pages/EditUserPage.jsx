import React, { useState } from "react";
import { useParams } from "react-router";
import {useLoading} from "../views/useLoading";
import {ErrorView} from "../views/ErrorView";
import {LoadingView} from "../views/LoadingView";
import {InputField} from "../views/InputField";

function EditUserForm({user, onSubmit}){
    const [firstName, setFirstName] = useState(user.firstName);
    const [lastName, setLastName] = useState(user.lastName);
    const [email, setEmail] = useState(user.email);

    async function submit(e){
        onSubmit(e, {firstName, lastName, email});
    }
    return (
        <form onSubmit={submit}>
            <InputField label={"First Name"} value={firstName} onChangeValue={setFirstName} />
            <InputField label={"Last Name"} value={lastName} onChangeValue={setLastName} />
            <InputField label={"Email"} value={email} onChangeValue={setEmail} />
            <button>Submit</button>
            <h4>Edit an existing user ({firstName} {lastName},{email})</h4>
        </form>
    );
}


export function EditUserPage({userApi}) {
    const {id} = useParams();
    const { data: user, loading, error, reload } = useLoading(
        async () => await userApi.getUser(id),
        [id]
    );
    async function handleSubmit(e, { firstName, lastName, email }) {
        e.preventDefault();
        await userApi.updateUser(id, {firstName, lastName, email});
    }

    if (error) {
        return <ErrorView error={error} reload={reload} />;
    }

    if (loading || !user) {
        return <LoadingView />;
    }

    return <EditUserForm user={user} onSubmit={handleSubmit}/>;
}
