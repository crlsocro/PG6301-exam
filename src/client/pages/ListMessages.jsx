import React from "react";
import { Link } from "react-router-dom";
import {useLoading} from "../views/useLoading";
import {ErrorView} from "../views/ErrorView";
import {LoadingView} from "../views/LoadingView";

export function ListMessages({messageApi}) {
    const { data: messages, error, loading, reload } = useLoading(
        async () => await messageApi.listMessage()
    );

    if (error) {
        return <ErrorView error={error} reload={reload} />;
    }

    if (loading || !messages) {
        return <LoadingView />;
    }

    return (
        <>
            <h1>List Messages</h1>
            {messages.map(({ id, text}) => (
                <li key={id}>
                    <Link to={`/messages/${id}`}>{text}</Link>
                </li>
            ))}
        </>
    );
}