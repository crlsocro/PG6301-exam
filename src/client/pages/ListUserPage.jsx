import React from "react";
import { Link } from "react-router-dom";
import {useLoading} from "../views/useLoading";
import {ErrorView} from "../views/ErrorView";
import {LoadingView} from "../views/LoadingView";



export function ListUserPage({userApi}) {
    const { data: users, error, loading, reload } = useLoading(
        async () => await userApi.listUsers()
    );

    if (error) {
        return <ErrorView error={error} reload={reload} />;
    }

    if (loading || !users) {
        return <LoadingView />;
    }

    return (
        <>
            <h1>List Users</h1>
            {users.map(({ id, firstName, lastName, email }) => (
                <li key={id}>
                    <Link to={`/users/${id}/edit`}>{firstName} {lastName}, {email}</Link>
                </li>
            ))}
        </>
    );
}