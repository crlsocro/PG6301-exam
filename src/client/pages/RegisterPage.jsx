import React, { useState } from "react";
import {InputField} from "../views/InputField";

export function RegisterPage({userApi}) {
    const [firstName, setFirstName] = useState("");
    const [lastName, setLastName] = useState("");
    const [email, setEmail] = useState("");

    async function submit(e) {
        e.preventDefault();
        await userApi.createUser({firstName, lastName, email});
    }
    return (
        <form onSubmit={submit}>
            <h4>Register a user: ({firstName} {lastName},{email})</h4>
            <br/>
            <InputField label={"First Name"} value={firstName} onChangeValue={setFirstName} />
            <InputField label={"Last Name"} value={lastName} onChangeValue={setLastName} />
            <InputField label={"Email"} value={email} onChangeValue={setEmail} />
            <button>Submit</button>
        </form>
    );
}