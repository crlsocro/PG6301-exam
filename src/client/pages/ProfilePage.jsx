import * as React from "react";
import { useLoading } from "../views/useLoading";
import {BrowserRouter, Link} from "react-router-dom";
import {Route, Switch} from "react-router";
import {ErrorView} from "../views/ErrorView";
import {LoadingView} from "../views/LoadingView";
import {ListUserPage} from "./ListUserPage";

export function ProfilePage({ loadProfile }) {
    const {loading, error, data} = useLoading(async () => await loadProfile());

    if (loading) {
        return <div>Loading...</div>;
    }
    if (error) {
        return (
            <div>
                <h1>An error occurred</h1>
                <div>{error.toString()}</div>
            </div>
        );
    }
//TODO: Legge til legge til bruker her
        return (
            <div>
                <h1>Profile</h1>
                <div>{data.name}</div>
                <div>{data.pid}</div>
            </div>

        );
}
