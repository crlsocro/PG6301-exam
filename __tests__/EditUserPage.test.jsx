import { EditUserPage } from "../src/client/pages/EditUserPage";
import * as ReactDOM from "react-dom";
import * as React from "react";
import { MemoryRouter } from "react-router";
import { act, Simulate } from "react-dom/test-utils";

async function renderForTest(child) {
    const container = document.createElement("div");
    await act(async () => {
        await ReactDOM.render(<MemoryRouter>{child}</MemoryRouter>, container);
    });
    return container;
}

describe("edit user page", () => {
    it("can show information about an existing user", async () => {
        const getUser = () => ({
            firstName: "Olaff",
            lastName: "Nordmannen",
            email: "olaff@nordmannen.no"
        });
        const container = await renderForTest(
            <EditUserPage userApi={{getUser}}/>
        );
        expect(container.innerHTML).toMatchSnapshot();
        expect(container.querySelector("h4").textContent).toEqual(
            "Edit an existing user (Olaff Nordmannen,olaff@nordmannen.no)"
        );
    });

    it("can show loading screen", async () => {
        const getUser = () => new Promise((resolve) => {});
        const container = await renderForTest(
            <EditUserPage userApi={{getUser}}/>
        );
        expect(container.innerHTML).toMatchSnapshot();
        expect(container.querySelector("div").textContent).toEqual("Loading ...");
    });

    it("can show error message", async () => {
        const getUser = () => {
            throw new Error("Failed to load");
        };
        const container = await renderForTest(
            <EditUserPage userApi={{getUser}}/>
        );
        expect(container.innerHTML).toMatchSnapshot();
        expect(container.querySelector("div").textContent).toEqual(
            "Something went wrong: Error: Failed to load"
        );
    });

    it("updates server on submit", async () => {
        const users = {
            firstName: "Olaffen",
            lastName: "Nordmanneni",
            email: "olaff@nordmannen.non"
        };
        const getUser = () => users;
        const updateUser = jest.fn();
        const container = await renderForTest(
            <EditUserPage userApi={{getUser, updateUser}}/>
        );
        Simulate.change(container.querySelector("input"), {
            target: { value: "Besumniki" },
        });
        Simulate.submit(container.querySelector("form"));
        expect(updateUser).toBeCalledWith(undefined, {
            ...users,
            firstName: "Besumniki",
        });
    });
});
