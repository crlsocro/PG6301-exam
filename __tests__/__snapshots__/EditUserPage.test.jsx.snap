// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`edit user page can show error message 1`] = `"<div>Something went wrong: Error: Failed to load</div><button>Try again</button>"`;

exports[`edit user page can show information about an existing user 1`] = `"<form><div><label>First Name: <input type=\\"text\\" value=\\"Olaff\\"></label></div><div><label>Last Name: <input type=\\"text\\" value=\\"Nordmannen\\"></label></div><div><label>Email: <input type=\\"text\\" value=\\"olaff@nordmannen.no\\"></label></div><button>Submit</button><h4>Edit an existing user (Olaff Nordmannen,olaff@nordmannen.no)</h4></form>"`;

exports[`edit user page can show loading screen 1`] = `"<div>Loading ...</div>"`;
